### ISSUE TITLE
Mapping JS objects to rendered SVG elements for easier theming

### ISSUE THREAD

**Issue by dswitzer** [amcharts/amcharts4#43]
One of the problems I find myself struggling with is how to customize an object (either via a theme or API calls). The reason I struggle is I'm often left having to guess at what property/class is generating the SVG element.

Is there any way that the SVG elements that are rendered could be mapped back to their JS class instances in any meaningful way? 

It would be great to be able to query the DOM for something like "[data-className='XYCursor'] to find instance of all the cursor elements. 

This would be really helpful when creating a theme, because I could inspect the DOM for hints as to what JS class I need to apply styles to.

**Comment by martynasma** [comment]
This is an excellent suggestion! Sorry for not getting back to you on this earlier.

I'm kinda reluctant to slap class names on every single element by default. However, we might discuss a setting which would enable it.

In the interim, each element in amCharts 4 has an undocument method `setClassName()`.

If you call it, it will slap a class name on it in the syntax `"amcharts-classnameinlowercase"`:

```
series.column.template.setClassName();
```

It does not take any parameters.

Meanwhile, we'll discuss if we can add a setting to auto-set class names.

**Comment by dswitzer** [comment]
Thanks for taking the time to reply.

Something like this is only needed for discovery when trying to style various aspects of the charts.

If there's something that could be added that wouldn't hurt performance when disabled, that would be really great. There'd be no reason to enable it outside of development. Maybe it would even alert/warn when the setting was on, to warn that is should be disabled.

The problem could also be resolved through better documentation and a really comprehensive theme guide.

My main pain point is it's just very difficult to figure out what I need to set in a theme to get the effect I'm after. I just thought if I could inspect the SVG elements being created for a hint of what created them, that would probably just help me out tremendously. Of course if everything would just end up say "sprite", that wouldn't be helpful. :)

**Comment by martynasma** [comment]
I have been reviewing this today. It turns out we already set class names that indicate what class was used for this element:

![image](https://user-images.githubusercontent.com/1458520/41472410-0a7d3844-70bf-11e8-97ef-17b1885af167.png)

I hope it's enough to identify class reference.

I'm going to shelve this for now, as it does not seems like a priority.

**Comment by dswitzer** [comment]
The problem is w/the existing classes, is they do not seem to map to the object names in any way that made sense to me. It was because of these classes I came up with the idea for this issue.

There's nothing in those class or attribute names, that I see, that I can use to map back to the objects passed into a theme for styling.

**Comment by Pauan** [comment]
Just FYI, the purpose of themes isn't really to set CSS styles or classes on the raw DOM nodes, instead it's to modify the amCharts objects:

```typescript
const theme: ITheme = (object: BaseObject) => {
    if (object instanceof Scrollbar) {
        object.background.fillOpacity = 0.4;
        object.thumb.background.fillOpacity = 0.5;
    }

    if (object instanceof AmChartsLogo) {
        object.urlTarget = "_self";
        object.align = "left";
        object.valign = "bottom";
    }
};

export default theme;
```

**Comment by dswitzer** [comment]
Correct. The problem I’m having when theming is knowing what class I need
to theme. I thought if I could inspect the SVG elements and find the
element I wanted and it had a hint like “Scrollbar”, then it would clue me
in to style the Scrollbar.

The biggest problem I have is I can find the SVG element, but then I have
trouble mapping it to the internal property/object used to create it.

**Comment by martynasma** [comment]
Gotcha. No, there's currently no intelligent relation between actual object class and SVG element.

Scrollbar you mentioned is a bunch of smaller elements, that each have own class that are reflected on classnames. But there's no classname that explicitly says "scrollbar".

I'm re-opening this so we can re-think this one more time.

**Comment by martynasma** [comment]
OK, so in dev version we now have this: (excerpt from changelog)

### Changed
- `class` attribute is no longer applied to elements by default. To enable it use new setting `am4core.options.autoSetClassName`.
- `class` attributes now will contain the whole inheritance chain, e.g. `"amcharts-Sprite amcharts-Container amcharts-Button"`. Class names are no longer lowercased.

### Added
- New global option `am4core.options.autoSetClassName` (default `false`). If set to `true` will set `class` attribute of all elements that reflect class element was created in, including inheritance, e.g. `"amcharts-Sprite amcharts-Container amcharts-Button"`.

### Removed
- Element-level `classNamePrefix` is no longer available. Use global `am4core.options.autoSetClassName` instead.

This will go into the next version we're going to release.

**Comment by dswitzer** [comment]
Thank you!

This seems like it would be very helpful.

**Comment by martynasma** [comment]
Implemented in 4.0.0-beta.30. ([changelog](https://github.com/amcharts/amcharts4/blob/master/dist/es2015/CHANGELOG.md#400-beta30---2018-06-22))

**Comment by dswitzer** [comment]
I haven't played with this too much yet, but this looks extremely helpful. 

Thank you!
